Use global cache directory to store shader cache

Calinou opened the pull request:
This has several upsides:

- `.godot/` is no longer bloated with shader cache, making ZIP archives of projects several megabytes larger when MRPs are uploaded to GitHub.
- Shader cache is shared across projects, which benefits startup times when you edit several projects.
- Shader cache is shared between the editor and running project, which benefits initial project startup times after opening the editor.

Given the potential implications of this change, I suggest not cherry-picking it until it's been thoroughly tested.

___

- This closes https://github.com/godotengine/godot/issues/73174.

Calinou committed:
Use global cache directory to store shader cache

This has several upsides:

- `.godot/` is no longer bloated with shader cache, making ZIP archives
  of projects several megabytes larger when MRPs are uploaded to GitHub.
- Shader cache is shared across projects, which benefits startup times
  when you edit several projects.
- Shader cache is shared between the editor and running project,
  which benefits initial project startup times after opening
  the editor.

oxeron commented:
Any chance this PR can be merged soon ?

akien-mga commented:
This requires discussion in a rendering team meeting probably. Shader cache is a priori dependent on the engine version, possibly what features are enabled in the project, etc., so making it global may not work that well in practice. Projects might break because they're using a shader cache that doesn't match their configuration, or the cache might end up being recreated every time you edit or run a new project.

I believe that the shader cache in `.godot` also gets included in exported projects so it's already pre-computed for players, which wouldn't be the case if it's global. I'm not 100% sure though, those assumptions need to be assessed.

Calinou commented:
> I believe that the shader cache in .godot also gets included in exported projects so it's already pre-computed for players, which wouldn't be the case if it's global.

It isn't included in the PCK, but this is what https://github.com/godotengine/godot/pull/102552 aims to address.

KoBeWi commented:
Note that each project has 2 shader caches - for editor inside `.godot` and for runtime inside user data folder. If they are the same, we could make editor use the user data cache.

Calinou commented:
> Note that each project has 2 shader caches - for editor inside `.godot` and for runtime inside user data folder. If they are the same, we could make editor use the user data cache.

I believe they are the same, although the editor can generate a few shaders specific to it (such as the 3D gizmo shaders and the animation onion skin shader). This shouldn't be a problem for the project though, as it won't use those extra files.

KoBeWi commented:
Yep, even if some shaders have different versions for the editor, a large part of the cache can be probably reused..

KoBeWi commented on editor/file_system/editor_paths.cpp:
```diff
@@ -250,10 +250,7 @@ EditorPaths::EditorPaths() {
 	// including shader cache subdir.
-	if (Engine::get_singleton()->is_project_manager_hint() || (Main::is_cmdline_tool() && !ProjectSettings::get_singleton()->is_project_loaded())) {
-		// Nothing to create, use shared editor data dir for shader cache.
-		Engine::get_singleton()->set_shader_cache_path(data_dir);
```
You could restore it and use `cache_dir`.

KoBeWi commented on editor/file_system/editor_paths.cpp:
```diff
@@ -287,4 +282,6 @@ EditorPaths::EditorPaths() {
 		}
 	}
+
+	Engine::get_singleton()->set_shader_cache_path(cache_dir);
```
Should be removed with the above changes.

KoBeWi reviewed (commented):
So I gave this a test again (idk if there were any changes since last time).

In summary, Godot actually has 3 shader cache folders:
- For Project Manager / headless engine, in the main data folder (with editor settings etc.)
- For editor, in the `.godot` folder
- For the running project, in the project's data directory (`user://`)

This PR changes the editor and project manager to share space in cache directory (the one with thumbnails and stuff). The running project still uses the project data directory.

I think the cache folder makes sense for project manager, but the editor could reuse project's shader cache (as they are going to have much in common, more than different projects). See my comments below for how to implement it.

Calinou committed:
[WIP] Use shared cache for editor/running project, keep separate cache for project manager

TODO:

- Fix `.godot/editor/` and `.godot/imported/` not being created,
  even if their respective `make_dir()` calls are replaced with `make_dir_recursive()`.
